### ISSUE TITLE
Production release crashes

### ISSUE THREAD

**Issue by Alamoz** [bitwalker/distillery#491]
Environment...

    Elixir 1.7.0 (compiled with Erlang/OTP 21)
    phoenix 1.3.0
    NodeJS version 8.10.0
    NPM version 3.5.2
    Operating system: Ubuntu (kenel version 4.15.0-32)

Expected behavior...

App runs without problems.
Actual behavior...

Phoenix app crashes. Erlang crash dump (erl_crash.dump) contains:

"Slogan: init terminating in do_boot ({cannot get bootfile,/home/phoenix/myapp/releases/0.0.1/config.boot})"

There is no config.boot file present in the release and there are other files not contained in the release tarball that exist in the release/x.x.x/ directory on the development machine.

**Comment by bitwalker** [comment]
Which version of Distillery? The latest release (2.0.2) fixes that issue.

**Comment by bitwalker** [comment]
I can't reproduce with 2.0.2 - can you clone `https://github.com/bitwalker/distillery-test` and update the deps to use `2.0.2` and repro with that? If you can, please give me the steps you followed to do so.

**Comment by Alamoz** [comment]
OK. Here are the results:

/tmp/test/releases/0.1.0+0$ ./bin/test start
[
  source: "/home/me/software/elixir/distillery-test/config/dev.exs",
  debug_level: {:on, [:passive]},
  env: nil
]
Hello from the post start!

-----------------------------------------------------------

/tmp/test/releases/0.1.0+0$ ./bin test ping
bash: ./bin: Is a directory

-----------------------------------------------------------

/tmp/test/releases/0.1.0+0$ ./bin/test remote_console
Erlang/OTP 21 [erts-10.0] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:1] [hipe]

Interactive Elixir (1.7.0) - press Ctrl+C to exit (type h() ENTER for help)
iex(test@127.0.0.1)1> GenServer.call(Test.Server, :ping)
:v1
iex(test@127.0.0.1)2> 

-----------------------------------------------------------

Here is the output of my /etc/systemd/system/myapp.service file:

[Unit]
Description=myapp
After=network.target

[Service]
WorkingDirectory=/home/phoenix/myapp
Environment=LANG=en_US.UTF-8
Environment=PORT=8888
Environment=MIX_ENV=prod
User=phoenix
RemainAfterExit=yes
Group=phoenix
Restart=on-failure

ExecStart=/home/phoenix/myapp/bin/sfapi start
ExecStop=/home/phoenix/myapp/bin/sfapi stop

[Install]
WantedBy=multi-user.target

**Comment by bitwalker** [comment]
Your second example was caused by a typo - you forgot the `/` between `./bin` and `test`.

It looks to me like everything is working - can you confirm that you actually are running 2.0.2 in your own app? If you run `grep 'distillery' mix.lock` you should see the SHA and version. Otherwise, you'll need to post your release config so I can see what settings may be causing the issue.

**Comment by Alamoz** [comment]
grep distillery mix.lock
  "distillery": {:hex, :distillery, "2.0.2", "58e15f6708e355b610a48356ef170e63ec7753e35e215b107e1e87ed79c65170", [:mix], [{:artificery, "~> 0.2", [hex: :artificery, repo: "hexpm", optional: false]}], "hexpm"},

**Comment by bitwalker** [comment]
@Alamoz What about your release config?

**Comment by bitwalker** [comment]
Duplicate of #489

**Comment by Alamoz** [comment]
There was no release config file. I just tried it with Distillery 2.0.3 and now it crashes with a vague Erlang error of:

Slogan: Kernel pid terminated (application_controller) ({application_start_failure,myapp,{{shutdown,{failed_to_start_child,'Elixir.MyapWeb.Endpoint',{shutdown,{failed_to_start_child,'Elixir.Phoenix.Endpoint.

I'm going to have to switch to Docker, I can't continuing having these types of problems. It was great using Distillery while it worked! Good luck!

**Comment by bitwalker** [comment]
How is there no release config file? I'm talking about `rel/config.exs`. Have you read the documentation on using Distillery?

In any case, I'm sorry to hear things are not working for you - I could have provided more help if you had provided the information requested in the template when you create a new issue, as well as details about how you are building and deploying the release artifact - the reason I request those things in the issue template to begin with is so that we don't have to waste time going back and forth.

FYI In your last message you pasted only part of the error, omitting the part which explains why the error occurred - the tail end of the structure printed contains important details.

**Comment by Alamoz** [comment]
Here is the top portion and last 999 lines of the Erlang crash dump file (erl_crash.dump):

=erl_crash_dump:0.5
Mon Aug 20 11:37:41 2018
Slogan: Kernel pid terminated (application_controller) ({application_start_failure,myapp,{{shutdown,{failed_to_start_child,'Elixir.myappWeb.Endpoint',{shutdown,{failed_to_start_child,'Elixir.Phoenix.Endpoint.
System version: Erlang/OTP 21 [erts-10.0] [source] [64-bit] [smp:1:1] [ds:1:1:10] [async-threads:1] [hipe]
Compiled: Tue Jun 19 18:01:11 2018
Taints: dyntrace,crypto,asn1rt_nif
Atoms: 33571
Calling Thread: scheduler:1
=scheduler:1
Scheduler Sleep Info Flags: 




set_funinfo_native_address
set_native_address
commit_patch_load
fun_to_address
merge_term
constants_size
alloc_data
enter_code
ref_set
ref_get
ref
array_update
array_sub
array_length
array
bitarray_update
bitarray_sub
bitarray
bytearray_update
bytearray_sub
bytearray
write_u32
write_u8
hipe_bifs
internal_select_delete
internal_delete_all
is_map_key
map_get
gather_carrier_info
gather_alloc_histograms
map_next
abort_connection
new_connection
get_dflags
iolist_to_iovec
set_signal
fmod
ceil
floor
has_prepared_code_on_load
copy_shared
size_shared
split
purge_module
check_dirty_process_code
is_process_executing_dirty
map_info
fun_info_mfa
take
cmp_term
values
update
remove
merge
keys
is_key
from_list
find
maps
map_size
is_map
inspect
printable_range
binary_to_float
float_to_binary
integer_to_binary
binary_to_integer
delete_element
insert_element
finish_loading
prepare_loading
dt_append_vm_tag_data
dt_prepend_vm_tag_data
dt_restore_tag
dt_spread_tag
dt_get_tag_data
dt_get_tag
dt_put_tag
posixtime_to_universaltime
universaltime_to_posixtime
check_old_code
native_name_encoding
file
is_translatable
internal_normalize_utf8
internal_native2name
internal_name2native
prim_file
nif_error
decode_unsigned
encode_unsigned
referenced_byte_size
list_to_bin
part
at
longest_common_suffix
longest_common_prefix
matches
compile_pattern
binary_part
finish_after_on_load
call_on_load_function
load_nif
setopts
give_away
dflag_unicode_io
binary_to_existing_atom
binary_to_atom
atom_to_binary
bin_is_7bit
characters_to_list
characters_to_binary
decode_packet
update_element
bitstring_to_list
list_to_bitstring
bit_size
byte_size
tuple_size
is_bitstring
list_to_existing_atom
iolist_to_binary
iolist_size
make_fun
string
is_boolean
get_module_info
warning_map
hibernate
is_module_native
make_stub_module
module_md5
get_chunk
lcnt_clear
lcnt_collect
lcnt_control
dirty
instructions
dist_ext_to_term
set_internal_state
get_internal_state
flat_size
same
disassemble
keyfind
keysearch
keymember
reverse
lists
run
format_error_int
loaded_drivers
try_unload
try_load
erl_ddll
perf_counter
getpid
list_env_vars
unset_env_var
set_env_var
get_env_var
os
match_spec_run_r
match_spec_compile
select_replace
select_reverse
select_count
select
update_counter
slot
safe_fixtable
rename
insert_new
insert
prev
next
member
match_object
last
lookup_element
lookup
is_compiled_ms
delete_object
delete
internal_request_all
match_spec_test
is_record
is_function
is_binary
is_reference
is_port
is_pid
is_number
is_integer
is_float
is_tuple
is_list
is_atom
subtract
'--'
append
'++'
'!'
is_builtin
get_stacktrace
raise
is_process_alive
demonitor
fun_to_list
port_to_list
ref_to_list
system_profile
system_monitor
system_info
system_flag
append_element
make_tuple
read_timer
cancel_timer
send_after
start_timer
pow
atan2
sqrt
log10
log2
log
exp
erfc
erf
atanh
atan
asinh
asin
acosh
acos
tanh
tan
sinh
sin
cosh
cos
math
bump_reductions
resume_process
suspend_process
seq_trace_print
seq_trace_info
seq_trace
trace_delivered
trace_info
trace_pattern
port_get_data
port_set_data
create_dist_channel
dirty_process_handle_signals
release_literal_area_switch
system_check
is_system_process
perf_counter_unit
time_unit
map_hashmap_children
term_type
map_to_tuple_keys
check_process_code
request_system_task
port_connect
port_close
port_control
port_command
port_call
port_info
dist_ctrl_get_data_notification
dist_ctrl_get_data
dist_ctrl_input_handler
dist_get_stat
setnode
spawn_opt
whereis
unlink
universaltime_to_localtime
universaltime
tuple_to_list
trunc
tl
time
term_to_binary
statistics
split_binary
spawn_link
spawn
setelement
self
round
registered
put
process_info
process_flag
pre_loaded
pid_to_list
open_port
system_time
monotonic_time
now
nodes
monitor_node
function_exported
module_loaded
md5_final
md5_update
md5_init
unique_integer
make_ref
localtime_to_universaltime
localtime
list_to_tuple
list_to_ref
list_to_port
list_to_pid
list_to_integer
list_to_float
list_to_binary
list_to_atom
link
length
is_alive
integer_to_list
hd
phash2
phash
halt
get_keys
get
garbage_collect
fun_info
float_to_list
float
external_size
exit_signal
erase
element
display_nl
display_string
display
delete_module
date
crc32_combine
crc32
binary_to_term
binary_to_list
atom_to_list
adler32_combine
adler32
abs
nifs
yield
yes
x86
xor
write_concurrency
wordsize
warning_msg
warning
wall_clock
waiting
visible
version
value
unload_cancelled
unloaded_only
unloaded
unless_suspending
uniq
unblock_normal
unblock
utf8
used
use_stdio
urun
unregister
unicode
ungreedy
undef
ucp
ucompile
type
try_clause
trim_all
trim
trap_exit
tracer
trace_status
trace_control_word
traced
trace
tpkt
total_run_queue_lengths_all
total_run_queue_lengths
total_heap_size
total_active_tasks_all
total_active_tasks
total
timestamp
'*'
timeout_value
time_offset
threads
thread_pool_size
this
term_to_binary_trap
table
system_architecture
system_version
system_limit
system_flag_scheduler_wall_time
system
suspending
suspended
suspend
sunrm
strict_monotonic_timestamp
strict_monotonic
stream
stop
stderr_to_stdout
status
start
stack_size
ssl_tls
spawned
spawn_driver
spawn_executable
size
silent
sigquit
sigtstp
sigsegv
sigint
sigstop
sigalrm
sigabrt
sigchld
sigill
sigusr2
sigusr1
sigterm
sighup
set_tcw_fake
set_tcw
set_seq_token
set_on_spawn
set_on_link
set_on_first_spawn
set_on_first_link
set_cpu_topology
set
serial
sequential_trace_token
sequential_tracer
sensitive
send_to_non_existing_process
send
seconds
second
scope
scientific
scheme
schedulers_online
scheduler_wall_time_all
scheduler_wall_time
scheduler_id
scheduler
sbct
save_calls
safe
runtime
running_procs
running_ports
running
runnable_procs
runnable_ports
runnable
run_queue_lengths_all
run_queue_lengths
run_queue
return_trace
return_to
return_from
restart
reset
report_errors
rem
reload
registered_name
register
refc
reductions
recent_size
receive
reason
ready_output
ready_input
read_concurrency
re_run_trap
re_pattern
re
raw
queue_size
quantify
purify
public
ptab_list_continue
protection
protected
profile
proc_sig
procs
process_dump
process_limit
process_display
process_count
processes_used
processes
process
private
priority
print
prepare_on_load
prepare
positive
port_op
port_limit
port_count
ports
port
pid
permanent
pending_reload
pending_purge_lambda
pending_process
pending_driver
pending
pause
'+'
parallelism
packet_size
packet
owner
overlapped_io
output
out_exiting
out_exited
out
os_version
os_type
os_pid
orelse
ordered_set
or
opt
open_error
open
on_load
on_heap
old_heap_size
old_heap_block_size
ok
offset
off_heap
nouse_stdio
notsup
notify
notempty_atstart
notempty
noteol
notbol
notalive
not_purged
not_pending
not_loaded_by_this_process
not_loaded
not_a_list
not
not_suspended
no_start_optimize
no_network
no_integer
no_float
nosuspend
noproc
noeol
nodeup
nodedown_reason
nodedown
node_type
node
noconnection
noconnect
no_auto_capture
none
nomatch
no
newline
new_uniq
new_ports
new_processes
new_index
new
never_utf
net_kernel_terminated
net_kernel
'/='
'=/='
need_gc
native_addresses
native
namelist
named_table
name
nanosecond
nano_seconds
multiline
multi_scheduling
more
monotonic_timestamp
monotonic
monitors
monitor_nodes
monitor
monitored_by
module_info
module
'-'
minor_version
minor
min_bin_vheap_size
min_heap_size
millisecond
milli_seconds
microstate_accounting
microsecond
micro_seconds
meta_match_spec
meta
merge_trap
messages
message_queue_len
message_queue_data
message
memory_types
memory_internal
memory
md5
mbuf_size
max_heap_size
maximum
max
match_spec_result
match_spec
match_limit_recursion
match_limit
match
major
magic_ref
machine
'<'
low
long_schedule
long_gc
logger
local
load_failure
load_cancelled
loaded
little
list_to_binary_continue
list
links
linked_in_driver
line_length
line_delimiter
line
lf
'=<'
ldflags
latin1
last_calls
large_heap
label
known
kill_ports
killed
kill
keypos
io
iterator
is_seq_trace
is_constant
invalid
instruction_counts
internal_error
internal
input
initial_call
init
info_msg
info
index
inconsistent
incomplete
inactive
in_exiting
in
ignore
if_clause
id
httph_bin
http_bin
http_error
http_eoh
http_header
http_request
http_response
https
httph
http
hipe_architecture
high
hide
hidden
heir
heart_port
heap_type
heap_sizes
heap_size
heap_block_size
have_dt_utag
handle
group_leader
grun
'>'
global
getting_unlinked
getting_linked
gather_system_check_result
gather_sched_wall_time_result
gather_microstate_accounting_result
gather_io_bytes
gather_gc_info_result
get_tcw
get_seq_token
generational
'>='
gc_minor_start
gc_minor_end
gc_max_heap_size
gc_major_start
gc_major_end
garbage_collection_info
garbage_collection
garbage_collecting
function_clause
functions
fullsweep_after
free
format_cpu_topology
force
flush_monitor_messages
flush
flags
firstline
first
fd
fcgi
extra
external
extended
exports
exiting
existing_ports
existing_processes
existing
exited
exit_status
exclusive
exception_trace
exception_from
exact_reductions
'ETS-TRANSFER'
ets
erts_internal
erts_dflags
erts_debug
erts_code_purger
error_logger
error_handler
erl_signal_server
erlang
erl_tracer
'=='
'=:='
eol
eof
env
endian
enabled
enable_trace
emulator
elib_malloc
einval
dupnames
duplicated
duplicate_bag
dsend_continue_trap
driver_options
driver
dotall
dollar_endonly
'$_'
'$$'
dmonitor_node
div
'/'
dist_data
dist_ctrl_put_data
dist_cmd
dist
display_items
discard
disabled
disable_trace
dirty_nif_finalizer
dirty_nif_exception
dirty_io
dirty_execution
dirty_cpu_schedulers_online
dirty_cpu
dirty_bif_trap
dirty_bif_result
dirty_bif_exception
dictionary
delay_trap
default
decimals
debug_flags
data
current_stacktrace
current_location
current_function
creation
crlf
cr
cpu_timestamp
cpu
counters
copy_literals
copy
control
context_switches
const
connection_closed
connected
connect
convert_time_unit
config_h
compressed
complete
compile
compat_rel
compact
commandv
command
code
closed
close
clock_service
clear
characters_to_list_int
characters_to_binary_int
'CHANGE'
cflags
cdr
cd
catchlevel
caseless
case_clause
capture
caller
call_time
call_count
busy_port
busy_dist_port
busy
build_type
bsr_unicode
bsr_anycrlf
bsr
bsl
breakpoint
break_ignored
bxor
bor
bnot
bm
blocked_normal
blocked
block_normal
block
binary_to_term_trap
binary_to_list_continue
binary_longest_suffix_trap
binary_longest_prefix_trap
binary_find_trap
binary_copy_trap
binary
bif_return_trap
big
band
bag
badsig
badmatch
badmap
badkey
badfun
badfile
badarity
badarith
badarg
backtrace_depth
backtrace
awaiting_unload
awaiting_load
await_sched_wall_time_modifications
await_result
await_proc_exit
await_port_send_result
await_microstate_accounting_modifications
await_exit
auto_connect
attributes
atom_used
atom
asynchronous
async
asn1
arity
arg0
args
apply
anycrlf
any
andthen
andalso
and
anchored
amd64
already_loaded
allow_passive_connect
alloc_util_allocators
allocator_sizes
allocator
allocated_areas
allocated
alloc_sizes
alloc_info
all_names
all_but_first
all
alive
active_tasks_all
active_tasks
active
ac
absoluteURI
abs_path
abort
'EXIT'
'UP'
'DOWN'
'3RT$'
undefined_lambda
undefined_function
nocatch
undefined
exit
error
throw
return
call
normal
timeout
infinity
''
'$end_of_table'
'nonode@nohost'
'_'
true
false
=end

**Comment by bitwalker** [comment]
The tail of the crash dump just contains the atoms created - you need the whole crash dump file to do anything meaningful with it, but it is best to upload those somewhere, as they are rather large.

In any case, troubleshooting your issue with the basics is very difficult - the following questions are what I typically ask everyone:

- Are you building and deploying on the same machine?
- If different, are they the same OS and architecture? Same kernel version isn't enough - Erlang is compiled against a variety of libraries, and if it is moved to another OS, those libraries may not be the correct versions, or not loadable. 
- Are you deploying a release with ERTS included (`include_erts: true` in the release config) or using a system-wide installation of Erlang? 
- Are all system dependencies present on the target (e.g. OpenSSL, etc.)?

Depending on the issue, the verbose logs from building the release can be important, or running the release with `DEBUG_BOOT=true` for verbose logs at boot. If you have a crash dump, that can be useful, but I'll typically request that you run [this script](https://github.com/ferd/recon/blob/master/script/erl_crashdump_analyzer.sh) against it yourself, and give me the results, rather than sending me the crash dump.

In your case, it _sounds_ like you are building and deploying to different machines, and that the target machine is not necessarily the same OS, but it's unclear; I also am not sure if you are including ERTS or not.

**Comment by Alamoz** [comment]
All those dependencies are satisfied. There is something else happening. One thing I did experience was a bunch of dependencies missing from re/config.exs but even after adding them the release still crashes. Strangely, I wasn't receiving the warning message yesterday but received it this morning:

==> Assembling release..
==> Building release myappweb:0.0.1 using environment prod
One or more direct or transitive dependencies are missing from
    :applications or :included_applications, they will not be included
    in the release:

        :cowboy
        :gettext
        :phoenix
        :phoenix_ecto
        :phoenix_html
        :phoenix_pubsub
        :postgrex

    This can cause your application to fail at runtime. If you are sure
    that this is not an issue, you may ignore this warning.

**Comment by Alamoz** [comment]
OK, the root cause of my problem (after your 2.0.3 release fixed the previous problem) was adding SSL support in the Phoenix app by un-commenting the SSL Support lines:

#config :myapp, MyappWeb.Endpoint,
# url: [host: "myapp.tld", port: 443],
# https: [:inet6,
#          port: 443,
#          keyfile: System.get_env("/etc/nginx/ssl/myapp.key"),
#          certfile: System.get_env("/etc/nginx/ssl/myapp.crt")]

I find it amusing how the Phoenix people reflexively assumed it was a Distillery problem, though there was indeed a problem with Distillery which you fixed. Too bad they are not interested in interoperating with Distillery, I like the concept of a self-contained release (I use "include_erts: true".)

Thank you for taking the time to offer suggestions.
